opencog/atoms/execution/EvaluationLink.cc

```diff
@@ -505,6 +507,29 @@ TruthValuePtr EvaluationLink::do_eval_scratch(AtomSpace* as,
 	{
 		return evelnk->getTruthValue();
 	}
+	else if (PREDICATE_FORMULA_LINK == t)
+	{
+		std::vector<double> nums;
```

Comment:
Should this block of code also be replaced by `do_formula()` call?

Reply:
I want to keep as-is; to avoid passing in an empty list of arguments

Reply:
1. It is not clear for me why list of arguments is empty here? Is it correct that `do_eval_scratch` is called only after variables substitution?
2. What is wrong with passing empty list of arguments if this will reduce the code bloat?

Reply:
1. Yes, `do_eval_scratch` is called after argument substitution and reduction has been made. (There is a mostly-finished "lazy execution" code path that is unused because GPN/GSN's don't know how to be lazy. Hmm. I should look at that again.

2. Yes, you are right. The two code paths are similar, one is longer and more complicated.  The decision to merge them seems very close to being a coin-flip; there are benefits and disadvantages, they seem to cancel each out out.